src/ImageArchive/ApiClient.php

```diff
@@ -0,0 +1,235 @@
+		'countperpage' => self::MEDIAS_PER_PAGE,
+		'format'       => 'json',
+		'pagenumber'   => 1,
+	];
+
+	/**
+	 * @var string The temporary authentication token.
+	 */
+	private $token;
+
+	/**
+	 * ApiClient constructor.
+	 *
+	 * @param string $token The temporary authentication token.
+	 */
+	private function __construct( string $token ) {
+		$this->token = $token;
+	}
+
+	/**
+	 * Use token from the cache to authenticate, or if none is present fetch a new one with the credentials
+	 * from the settings.
+	 *
+	 * @return static Authenticated instance.
+	 * @throws RemoteCallFailed Authentication failed.
+	 */
+	public static function from_cache_or_credentials(): self {
+		$cached_token = get_transient( self::TOKEN_CACHE_KEY );
+
+		if ( false !== $cached_token ) {
+			return new self( $cached_token );
+		}
+
+		return self::from_settings();
+	}
+
+	/**
+	 * Authenticate with the credentials from the settings.
+	 *
+	 * @return static Authenticated instance.
+	 * @throws RemoteCallFailed Authentication failed.
+	 */
+	public static function from_settings(): self {
+		$p4ml_settings = get_option( 'p4ml_main_settings' );
+
+		return self::from_credentials( $p4ml_settings['p4ml_api_username'], $p4ml_settings['p4ml_api_password'] );
+	}
+
+	/**
+	 * Fetch a token with provided credentials and use it.
+	 *
+	 * @param string $username The username of the API account.
+	 * @param string $password The password of the API account.
+	 *
+	 * @return static Authenticated instance.
+	 * @throws RemoteCallFailed Authentication failed.
+	 */
+	public static function from_credentials( string $username, string $password ): self {
+		$token = self::fetch_token( $username, $password );
+
+		return new self( $token );
+	}
+
+	/**
+	 * Call the authentication endpoint with credentials to fetch a token.
+	 *
+	 * @param string $username The username of the API account.
+	 * @param string $password The password of the API account.
+	 *
+	 * @return string The authentication token.
+	 * @throws RemoteCallFailed Authentication failed.
+	 */
+	private static function fetch_token( string $username, string $password ): string {
+		$response = wp_safe_remote_post(
+			self::AUTH_URL,
+			[
+				'body'    => [
+					'Login'    => $username,
+					'Password' => $password,
+					'format'   => 'json',
+				],
+				'timeout' => self::RESPONSE_TIMEOUT,
+			]
+		);
+		// Authentication failure.
+		if ( is_wp_error( $response ) ) {
+			$response = $response->get_error_message() . ' ' . $response->get_error_code();
+
+		} elseif ( WP_Http::ACCEPTED !== $response['response']['code'] ) {
+			$response = $response['response']['message'] . ' ' . $response['response']['code'];
+		}
+
+		if ( ! is_array( $response ) || ! ( $response['body'] ) ) {
```

Comment:
`wp_safe_remote_post` says that it will return an array `containing 'headers', 'body', 'response', 'cookies', and 'filename' elements` but I can't find any guarantee about that in wp code, when returning things from `apply_filter` hooks. So maybe `if ( ! is_array( $response ) || empty( $response['body'] ) ) {` would be safer (to no get a notice) ?

Reply:
Good point, in fact `$response` could be anything as it's passed through a filter before it's returned https://github.com/WordPress/WordPress/blob/master/wp-includes/class-http.php#L446